pkg/controller/clusters/status.go

```diff
@@ -22,10 +22,61 @@ func UpdateClusterStatusForControlPlane(ctx context.Context, client client.Clien
 	}
 
 	updateControlPlaneInitializedCondition(cluster, kcp)
+	updateControlPlaneReadyCondition(cluster, kcp)
 
 	return nil
 }
 
+// updateControlPlaneReadyCondition updates the ControlPlaneReady condition, after checking the state of the control plane
+// in the cluster.
+func updateControlPlaneReadyCondition(cluster *anywherev1.Cluster, kcp *controlplanev1.KubeadmControlPlane) {
+	initializedCondition := conditions.Get(cluster, anywherev1.ControlPlaneInitializedCondition)
+	if initializedCondition.Status != "True" {
+		conditions.MarkFalse(cluster, anywherev1.ControlPlaneReadyCondition, initializedCondition.Reason, initializedCondition.Severity, initializedCondition.Message)
+		return
+	}
+
+	// We make sure to check that the status is up to date before using it
+	if kcp.Status.ObservedGeneration != kcp.ObjectMeta.Generation {
+		conditions.MarkFalse(cluster, anywherev1.ControlPlaneReadyCondition, anywherev1.OutdatedInformationReason, clusterv1.ConditionSeverityInfo, "")
+		return
+	}
+
+	// The control plane should be marked ready when the count specified in the spec is
+	// equal to the ready number of nodes in the cluster and they're all of the right version specified.
+
+	expected := cluster.Spec.ControlPlaneConfiguration.Count
+	totalReplicas := int(kcp.Status.Replicas)
+
+	// First, in the case of a rolling upgrade, we get the number of outdated nodes, and as long as there are some,
+	// we want to reflect in the message that the Cluster is in progres upgdating the old nodes with the
+	// the new machine spec.
+	updatedReplicas := int(kcp.Status.UpdatedReplicas)
+	totalOutdated := totalReplicas - updatedReplicas
+
+	if totalOutdated > 0 {
+		conditions.MarkFalse(cluster, anywherev1.ControlPlaneReadyCondition, anywherev1.RollingUpgradeInProgress, clusterv1.ConditionSeverityInfo, "Control plane nodes not up-to-date yet, %d rolling (%d up to date)", totalReplicas, updatedReplicas)
+		return
+	}
+
+	// Then, we check that the number of nodes in the cluster match the expected amount. If not, we
+	// mark that the Cluster is scaling up or scale down the control plane replicas to the expected amount.
+	if totalReplicas < expected {
+		conditions.MarkFalse(cluster, anywherev1.ControlPlaneReadyCondition, anywherev1.ScalingUpReason, clusterv1.ConditionSeverityInfo, "Scaling up control plane nodes, %d expected (%d actual)", expected, totalReplicas)
+		return
+	}
+
+	if totalReplicas > expected {
+		conditions.MarkFalse(cluster, anywherev1.ControlPlaneReadyCondition, anywherev1.ScalingDownReason, clusterv1.ConditionSeverityInfo, "Scaling down control plane nodes, %d expected (%d actual)", expected, totalReplicas)
+		return
+	}
+
+	readyReplicas := int(kcp.Status.ReadyReplicas)
+	if readyReplicas == expected {
```

Comment:
what happens when this is not true?

Reply:
If it isn't true, then no update is made. The intention here was to preserve the previous false status until everything is ready.

Though, I suppose we may have a situation where there is no previous status, and the nodes aren't ready yet. I think then we'll need another "reason" here. Something like `ProvisioningInProgress`?

Reply:
Under what scenarios we might get here?
- no outdated replicas. This means all replicas existing are at the desired version.
- number of existing replicas is the same as desired replicas.

So we are either:
- provisioning the last replica of the CP (machine already created but not healthy yet)
- one or more of the existing machines became not ready

So `ProvisioningInProgress` only covers half of those scenarios. What about "NodesNotReady"? that is true for both scenarios.
And in the future we can always add logic to distinguish between the two scenarios if necessary.

Reply:
Yeah `NodesNotReady` sounds good to me in that case